src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+													useIndex, NULL, scanKeyCount,
+													scanKey);
+
+	HeapTuple heapTuple = systable_getnext(scanDescriptor);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		Form_pg_index indexForm = (Form_pg_index) GETSTRUCT(heapTuple);
+
+		if (indexForm->indisprimary || indexForm->indisunique ||
+			indexForm->indisexclusion)
+		{
+			/*
+			 * Skip the indexes that are not implied by explicitly executing
+			 * a CREATE INDEX command.
+			 */
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
+
+		Oid indexId = indexForm->indexrelid;
+
+		char *indexName = get_rel_name(indexId);
+
+		indexNameList = lappend(indexNameList, pstrdup(indexName));
+
+		heapTuple = systable_getnext(scanDescriptor);
+	}
+
+	systable_endscan(scanDescriptor);
+	heap_close(pgIndex, AccessShareLock);
+
+	/* revert back to original search_path */
+	PopOverrideSearchPath();
+
+	return indexNameList;
+}
+
+
+/*
+ * CreateShellTableForCitusLocalTable creates the shell table for the citus
+ * local tables by executing the given commands necessary to recreate the
+ * table.
+ */
+static void
+CreateShellTableForCitusLocalTable(List *shellTableDDLEvents)
+{
+	Assert(list_length(shellTableDDLEvents) > 0);
+
+	char *ddlCommand = NULL;
+	foreach_ptr(ddlCommand, shellTableDDLEvents)
+	{
+		StringInfo semicolonEndedCommand = makeStringInfo();
+		appendStringInfo(semicolonEndedCommand, "%s;", ddlCommand);
+
+		const char *commandString = semicolonEndedCommand->data;
+
+		Node *parseTree = ParseTreeNode(commandString);
+
+		/*
+		 * If the command defines a constraint, initially do not validate it
+		 * as shell table has no data in it. Note that this is only needed
+		 * before inserting metadata for citus local table. This is because,
+		 * after that, we already won't process constraints on the shell table.
+		 */
+		if (IsA(parseTree, AlterTableStmt))
+		{
+			AlterTableStmt *alterTableStmt = castNode(AlterTableStmt, parseTree);
+
+			/* we create ALTER TABLE commands so they only have one subcommand */
+			Assert(list_length(alterTableStmt->cmds) == 1);
+			AlterTableCmd *alterTableCmd =
+				castNode(AlterTableCmd, linitial(alterTableStmt->cmds));
+
+			if (alterTableCmd->subtype == AT_AddConstraint)
+			{
+				Constraint *constraint = (Constraint *) alterTableCmd->def;
+				constraint->skip_validation = true;
+			}
+		}
+
+		CitusProcessUtility(parseTree, commandString, PROCESS_UTILITY_TOPLEVEL,
+							NULL, None_Receiver, NULL);
+	}
+}
+
+
+/*
+ * InsertMetadataForCitusLocalTable inserts necessary metadata for the citus
+ * local table to the following metadata tables:
+ * pg_dist_partition, pg_dist_shard & pg_dist_placement.
+ */
+static void
+InsertMetadataForCitusLocalTable(Oid citusLocalTableId, uint64 shardId)
+{
+	Assert(OidIsValid(citusLocalTableId));
+	Assert(shardId != INVALID_SHARD_ID);
+
+	char distributionMethod = DISTRIBUTE_BY_NONE;
+	char replicationModel = ReplicationModel;
+
+	Assert(replicationModel != REPLICATION_MODEL_2PC);
```

Comment:
> Do we give an error if replication model is "statement"?

No, it is also allowed.

> also how can this be set to REPLICATION_MODEL_2PC? As far as I can see these are the available options for ReplicationModel:

(Not much possible but) In case someone mistakenly enables `2PC` for `replication_model_options` during development, we should point that citus local tables shouldn't have `2PC` as the only thing makes the difference from reference tables is the replication model.